invenio_pidstore/providers/datacite.py

```diff
@@ -31,20 +38,71 @@ class DataCiteProvider(BaseProvider):
     default_status = PIDStatus.NEW
     """Default status for newly created PIDs by this provider."""
 
+    doi_prefix_regexp = re.compile(
+        r"10\.\d+(\.\d+)*$"
+    )
+
+    @classmethod
+    def valid_doi_prefix(cls, prefix):
+        """Matches if prefix is a DOI prefix.
+
+        Potential TODO: add to idutils module.
+        """
+        return cls.doi_prefix_regexp.match(prefix)
+
+    @classmethod
+    def generate_doi(cls, prefix, suffix_length=10, split_every=5,
+                     checksum=True):
+        """Generate random DOI with `prefix`."""
+        if not cls.valid_doi_prefix(prefix):
+            logger.exception("Invalid DOI prefix", extra={'prefix': prefix})
+            return None
+
+        if checksum and suffix_length < 3:
+            logger.exception(
+                "Invalid suffix_length. At least 3 if checksum enabled",
+                extra={'suffix_length': suffix_length}
+            )
+            return None
+
+        generator = random.SystemRandom()
+        length = suffix_length - 2 if checksum else suffix_length
+        number = generator.getrandbits(length * 5)
+
+        return (
+            prefix +
+            "/" +
+            base32.encode(
+                number,
+                split_every=split_every,
+                checksum=checksum
+            )
+        )
+
     @classmethod
-    def create(cls, pid_value, **kwargs):
+    def create(cls, pid_value=None, prefix=None, **kwargs):
         """Create a new record identifier.
 
         For more information about parameters,
         see :meth:`invenio_pidstore.providers.BaseProvider.create`.
 
         :param pid_value: Persistent identifier value.
+        :param prefix: DOI prefix if pid_value is None
         :params **kwargs: See
             :meth:`invenio_pidstore.providers.base.BaseProvider.create` extra
             parameters.
         :returns: A :class:`invenio_pidstore.providers.DataCiteProvider`
             instance.
         """
+        if pid_value is None:
+            prefix = (
+                prefix or
+                current_app.config.get('PIDSTORE_DATACITE_DOI_PREFIX')
+            )
+            pid_value = cls.generate_doi(prefix)
```

Comment:
**Moderate**I agree on this. Whats the overhead on querying db every creation?

Reply:
It is definitely a performance/convenience tradeoff (but I would argue it is NOT an integrity one). On one hand, generation is faster and has no DB bottleneck (performant!). On the other, it does not have full 100% guarantee that the returned DOI is unique. This would cause an exception and the need to resubmit a request (inconvenient!). 

BUT :smile: 

Since `PersistentIdentifier` has a primary key on DOI value, creating a `PersistentIdentifier` with a non-unique DOI is not possible. Like mentioned above, an exception will be raised and so there is no integrity problem.

Also, although not 100%, the guarantee of uniqueness (no collision) is pretty high. For length=8, there are (2^5)^8 ~= 10^12 unique ids. A collision is something like 1/ 10^12 * #previously_generated / 10^12 i.e. in the quadrillionth chance of happening i.e. the guarantee is 99.9999999....% of uniqueness. And we can just up the suffix length if we want :)       

Anyway, the point is well taken that a guarantee would be great. If this is a tradeoff people want, I can change it as suggested. It would be interesting to hear people's thought.

Reply:
:smirk_cat: and it seems like performance is an issue in this module: https://github.com/inveniosoftware/invenio-pidstore/issues/32 (I am partially trolling haha)

Reply:
Now that this is gonna be moved to an external repo, the check should definitely have to be made outside this function (e.g. as @fenekku mentioned, catching the exception on the `PersistentIdentifier`)